### PR TITLE
Task04 Валерий Мацкевич HSE

### PR DESCRIPTION
**Pull request by blonded04**
<details><summary>Локальный вывод</summary><p>

<pre>
$ ./build/matrix_transpose 
OpenCL devices:
  Device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Using device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Data generated for M=4096, K=4096
[matrix_transpose_naive]
    GPU: 0.0169246+-0.000363228 s
    GPU: 991.289 millions/s
[matrix_transpose_local_bad_banks]
    GPU: 0.0162895+-3.76592e-05 s
    GPU: 1029.94 millions/s
[matrix_transpose_local_good_banks]
    GPU: 0.0162737+-4.11965e-05 s
    GPU: 1030.94 millions/s
$ ./build/matrix_multiplication 
OpenCL devices:
  Device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Using device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Data generated for M=1024, K=1024, N=1024
CPU: 10.3802+-0 s
CPU: 0.192675 GFlops
[naive, ts=4]
    GPU: 0.929132+-0.00651321 s
    GPU: 2.15255 GFlops
    Average difference: 0.000149043%
[naive, ts=8]
    GPU: 0.939817+-0.00427403 s
    GPU: 2.12807 GFlops
    Average difference: 0.000149043%
[naive, ts=16]
    GPU: 1.20814+-0.00336305 s
    GPU: 1.65544 GFlops
    Average difference: 0.000149043%
[local, ts=4]
    GPU: 0.65125+-0.306172 s
    GPU: 3.07102 GFlops
    Average difference: 0.000149043%
[local, ts=8]
    GPU: 0.564746+-0.284267 s
    GPU: 3.54142 GFlops
    Average difference: 0.000149043%
[local, ts=16]
    GPU: 0.310659+-0.00316622 s
    GPU: 6.43794 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=2]
    GPU: 1.23883+-0.633773 s
    GPU: 1.61442 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=4]
    GPU: 0.543547+-0.0472318 s
    GPU: 3.67953 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=2]
    GPU: 0.59599+-0.21084 s
    GPU: 3.35576 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=4]
    GPU: 0.671404+-0.117922 s
    GPU: 2.97883 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=8]
    GPU: 0.524906+-0.18261 s
    GPU: 3.81021 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=2]
    GPU: 0.391485+-0.0641349 s
    GPU: 5.10876 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=4]
    GPU: 0.445875+-0.146809 s
    GPU: 4.48556 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=8]
    GPU: 0.498357+-0.0494717 s
    GPU: 4.01319 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=16]
    GPU: 0.361147+-0.04432 s
    GPU: 5.53792 GFlops
    Average difference: 0.000149043%
</pre>

</p></details>

<details><summary>Вывод Github CI</summary><p>

<pre>
$ ./matrix_transpose
OpenCL devices:
  Device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Using device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Data generated for M=4096, K=4096
[matrix_transpose_naive]
    GPU: 0.0152054+-0.000145498 s
    GPU: 1103.37 millions/s
[matrix_transpose_local_bad_banks]
    GPU: 0.0225613+-0.000898613 s
    GPU: 743.627 millions/s
[matrix_transpose_local_good_banks]
    GPU: 0.0224438+-0.000941109 s
    GPU: 747.522 millions/s
$ ./matrix_multiplication
OpenCL devices:
  Device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Using device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Data generated for M=1024, K=1024, N=1024
CPU: 6.33776+-0 s
CPU: 0.315569 GFlops
[naive, ts=4]
    GPU: 0.368016+-0.00387518 s
    GPU: 5.43454 GFlops
    Average difference: 0.000149043%
[naive, ts=8]
    GPU: 0.373403+-0.00552401 s
    GPU: 5.35614 GFlops
    Average difference: 0.000149043%
[naive, ts=16]
    GPU: 0.372796+-0.00299166 s
    GPU: 5.36486 GFlops
    Average difference: 0.000149043%
[local, ts=4]
    GPU: 0.536499+-0.000822247 s
    GPU: 3.72787 GFlops
    Average difference: 0.000149043%
[local, ts=8]
    GPU: 0.310242+-0.00216126 s
    GPU: 6.44658 GFlops
    Average difference: 0.000149043%
[local, ts=16]
    GPU: 0.2867+-0.00103005 s
    GPU: 6.97592 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=2]
    GPU: 0.514041+-0.00417185 s
    GPU: 3.89074 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=4]
    GPU: 0.441325+-0.00117908 s
    GPU: 4.53181 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=2]
    GPU: 0.318689+-0.00104722 s
    GPU: 6.2757 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=4]
    GPU: 0.284307+-0.00114197 s
    GPU: 7.03465 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=8]
    GPU: 0.275566+-0.00210648 s
    GPU: 7.25779 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=2]
    GPU: 0.239857+-0.00118266 s
    GPU: 8.33831 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=4]
    GPU: 0.227042+-0.00149647 s
    GPU: 8.80894 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=8]
    GPU: 0.199619+-0.00168463 s
    GPU: 10.0191 GFlops
    Average difference: 0.000149043%
</pre>

</p></details>

### CODE REVIEW

File: src/cl/matrix_transpose.cl [hunk 1]
```diff
@@ -5,17 +5,48 @@
...
+    unsigned int width, unsigned int height)
 {
-    // TODO
+    const int gid_x = get_global_id(0);
+    const int gid_y = get_global_id(1);
+    const int lid_x = get_local_id(0);
+    const int lid_y = get_local_id(1);
+
+    __local float buffer[TILE_SIZE][TILE_SIZE];
+    buffer[lid_x][lid_y] = matrix[gid_x * height + gid_y];
```
Comment (simiyutin) [inline src/cl/matrix_transpose.cl:30]: Не коалесд доступ к глобальной памяти + в таком использовании локального массива нет смысла, поток использует то же значение что сам и загрузил

File: src/cl/matrix_transpose.cl [hunk 2]
```diff
@@ -5,17 +5,48 @@
...
+    unsigned int width, unsigned int height)
 {
-    // TODO
+    const int gid_x = get_global_id(0);
+    const int gid_y = get_global_id(1);
+    const int lid_x = get_local_id(0);
+    const int lid_y = get_local_id(1);
+
+    __local float buffer[TILE_SIZE][TILE_SIZE + 1];
+    buffer[lid_y][lid_x] = matrix[gid_x * height + gid_y];
```
Comment (simiyutin) [inline src/cl/matrix_transpose.cl:47]: Не коалесд доступ к глобальной памяти + в таком использовании локального массива нет смысла, поток использует то же значение что сам и загрузил

File: src/cl/matrix_multiplication.cl [hunk 3]
```diff
@@ -7,21 +7,95 @@
...
+    const int gid_y = get_global_id(1);
+    const int lid_x = get_local_id(0);
+    const int lid_y = get_local_id(1);
+
+    __local float lhs_tile[TILE_SIZE][TILE_SIZE];
+    __local float rhs_tile[TILE_SIZE][TILE_SIZE];
+
+    float sum = 0;
+    for (int tile = 0; tile < k; tile += TILE_SIZE) {
+        lhs_tile[lid_x][lid_y] = lhs[gid_x * k + lid_y + tile];
```
Comment (simiyutin) [inline src/cl/matrix_multiplication.cl:44]: не коалесд доступ

File: src/cl/matrix_multiplication.cl [hunk 4]
```diff
@@ -7,21 +7,95 @@
...
-    // TODO
+    const int gid_x = get_global_id(0);
+    const int gid_y = get_global_id(1);
+    const int lid_x = get_local_id(0);
+    const int lid_y = get_local_id(1);
+
+    __local float lhs_tile[TILE_SIZE][TILE_SIZE];
+    __local float rhs_tile[TILE_SIZE][TILE_SIZE];
+
+    // for some reason, `sum[WORK_PER_THREAD] = { 0 }` doesn't work (sic!):
```
Comment (simiyutin) [inline src/cl/matrix_multiplication.cl:75]: :scream:

File: src/cl/matrix_multiplication.cl [hunk 5]
```diff
@@ -7,21 +7,95 @@
...
+    float sum[WORK_PER_THREAD];
+    for (int i = 0; i < WORK_PER_THREAD; i++) {
+        sum[i] = 0;
+    }
+    for (int tile = 0; tile < k; tile += TILE_SIZE) {
+        for (int i = 0; i < WORK_PER_THREAD; i++) {
```
Comment (simiyutin) [inline src/cl/matrix_multiplication.cl:81]: не коалесд доступ

File: src/main_matrix_multiplication.cpp [hunk 6]
```diff
@@ -48,47 +48,41 @@ struct KernelConfig {
...
     return KernelConfig{kernel_name, work_size, defines, prefix};
 }
 
-KernelConfig makeLocalWPTConfig(unsigned int tile_size, unsigned int wpt)
-{
-    throw std::runtime_error("not implemented");
+KernelConfig makeLocalWPTConfig(unsigned int tile_size, unsigned int wpt) {
     std::string kernel_name = "matrix_multiplication_local_wpt";
-    gpu::WorkSize work_size(0, 0/*TODO*/);
+    gpu::WorkSize work_size(tile_size / wpt, tile_size, M / wpt, N);
```
Comment (simiyutin) [inline src/main_matrix_multiplication.cpp:69]: сжатие рабочего пространства и рабочей группы по строчкам, что приводит к некоалесности по причине, которую обсуждали на лекции:
Screenshot from 2024-10-12 21-52-31

**Comment by blonded04** [comment]
Всё исправил, что-то я поплыл с коалесцностью на этой домашке:
<details><summary>Локальный вывод</summary><p>

<pre>
$ ./build/matrix_transpose 
OpenCL devices:
  Device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Using device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Data generated for M=4096, K=4096
[matrix_transpose_naive]
    GPU: 0.0157434+-3.5513e-05 s
    GPU: 1065.67 millions/s
[matrix_transpose_local_bad_banks]
    GPU: 0.0164779+-6.62219e-05 s
    GPU: 1018.16 millions/s
[matrix_transpose_local_good_banks]
    GPU: 0.0194112+-0.000115898 s
    GPU: 864.304 millions/s
$ ./build/matrix_multiplication 
OpenCL devices:
  Device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Using device #0: CPU. pthread-AMD Ryzen 5 5500U with Radeon Graphics. AuthenticAMD. Total memory: 13334 Mb
Data generated for M=1024, K=1024, N=1024
CPU: 9.26803+-0 s
CPU: 0.215796 GFlops
[naive, ts=4]
    GPU: 0.651386+-0.00693359 s
    GPU: 3.07038 GFlops
    Average difference: 0.000149043%
[naive, ts=8]
    GPU: 1.24071+-0.00616693 s
    GPU: 1.61198 GFlops
    Average difference: 0.000149043%
[naive, ts=16]
    GPU: 1.27211+-0.00265694 s
    GPU: 1.57219 GFlops
    Average difference: 0.000149043%
[local, ts=4]
    GPU: 0.730875+-0.253384 s
    GPU: 2.73644 GFlops
    Average difference: 0.000149043%
[local, ts=8]
    GPU: 0.446461+-0.00338884 s
    GPU: 4.47968 GFlops
    Average difference: 0.000149043%
[local, ts=16]
    GPU: 0.479548+-0.278681 s
    GPU: 4.17059 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=2]
    GPU: 0.940105+-0.418283 s
    GPU: 2.12742 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=4]
    GPU: 0.575348+-0.0166406 s
    GPU: 3.47616 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=2]
    GPU: 0.640607+-0.207482 s
    GPU: 3.12204 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=4]
    GPU: 0.789098+-0.657385 s
    GPU: 2.53454 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=8]
    GPU: 0.601464+-0.328727 s
    GPU: 3.32522 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=2]
    GPU: 0.450417+-0.164779 s
    GPU: 4.44032 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=4]
    GPU: 0.36913+-0.00197336 s
    GPU: 5.41814 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=8]
    GPU: 0.516852+-0.326636 s
    GPU: 3.86958 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=16]
    GPU: 0.618173+-0.352849 s
    GPU: 3.23534 GFlops
    Average difference: 0.000149043%
</pre>

</p></details>

<details><summary>Вывод Github CI</summary><p>

<pre>
$ ./matrix_transpose
OpenCL devices:
  Device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Using device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Data generated for M=4096, K=4096
[matrix_transpose_naive]
    GPU: 0.0150901+-8.39546e-05 s
    GPU: 1111.8 millions/s
[matrix_transpose_local_bad_banks]
    GPU: 0.013398+-4.4169e-05 s
    GPU: 1252.22 millions/s
[matrix_transpose_local_good_banks]
    GPU: 0.0123532+-3.18931e-05 s
    GPU: 1358.13 millions/s
$ ./matrix_multiplication
OpenCL devices:
  Device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Using device #0: CPU. AMD EPYC 7763 64-Core Processor                . Intel(R) Corporation. Total memory: 15991 Mb
Data generated for M=1024, K=1024, N=1024
CPU: 6.33898+-0 s
CPU: 0.315508 GFlops
[naive, ts=4]
    GPU: 0.297196+-0.00687361 s
    GPU: 6.72956 GFlops
    Average difference: 0.000149043%
[naive, ts=8]
    GPU: 0.345108+-0.00716501 s
    GPU: 5.79528 GFlops
    Average difference: 0.000149043%
[naive, ts=16]
    GPU: 0.35401+-0.0125112 s
    GPU: 5.64956 GFlops
    Average difference: 0.000149043%
[local, ts=4]
    GPU: 0.598741+-0.000859581 s
    GPU: 3.34034 GFlops
    Average difference: 0.000149043%
[local, ts=8]
    GPU: 0.151959+-0.00045314 s
    GPU: 13.1614 GFlops
    Average difference: 0.000149043%
[local, ts=16]
    GPU: 0.097215+-0.000384764 s
    GPU: 20.573 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=2]
    GPU: 0.522618+-0.00120929 s
    GPU: 3.82689 GFlops
    Average difference: 0.000149043%
[local wpt, ts=4, wpt=4]
    GPU: 0.43653+-0.00185709 s
    GPU: 4.58159 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=2]
    GPU: 0.143992+-0.000368186 s
    GPU: 13.8896 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=4]
    GPU: 0.147978+-0.000263492 s
    GPU: 13.5155 GFlops
    Average difference: 0.000149043%
[local wpt, ts=8, wpt=8]
    GPU: 0.135031+-0.000365764 s
    GPU: 14.8114 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=2]
    GPU: 0.0832022+-0.000359271 s
    GPU: 24.0378 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=4]
    GPU: 0.0834775+-0.000358338 s
    GPU: 23.9586 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=8]
    GPU: 0.0838542+-0.000326071 s
    GPU: 23.8509 GFlops
    Average difference: 0.000149043%
[local wpt, ts=16, wpt=16]
    GPU: 0.0826043+-0.000779651 s
    GPU: 24.2118 GFlops
    Average difference: 0.000149043%
</pre>

</p></details>

**Comment by simiyutin** [comment]
> Всё исправил, что-то я поплыл с коалесцностью на этой домашке:

Да, в этой домашке легко с ней поплыть)

**Comment by simiyutin** [comment]
Все хорошо, задача зачтена, 10/10 баллов :+1:

**Pull request closed without merging**